EventAggregator/EventAggregator.cs

```diff
@@ -424,20 +544,109 @@ public bool HandlesMessage(object message)
                 return handled;
             }
 
-            public void TryHandle<TListener>(object target, object message, out bool wasHandled)
+            public bool TryHandle<TListener>(object target, object message)
                 where TListener : class
             {
-                wasHandled = false;
                 if (target == null)
                 {
-                    return;
+                    return false;
                 }
 
-                if (!Handles<TListener>() && !HandlesMessage(message)) return;
-
+                if (!Handles<TListener>() && !HandlesMessage(message)) return false;
+#if !ASYNC && !UNWRAP_EX
                 _handlerMethod.Invoke(target, new[] {message});
-                wasHandled = true;
+#elif ASYNC && !UNWRAP_EX
+                if (TypeHelper.IsAssignableToGenericType(target.GetType(), typeof(IListenerAsync<>)))
+                {
+                    ((dynamic) _handlerMethod.Invoke(target, new[] {message})).Wait();
+                }
+                else
+                {
+                    _handlerMethod.Invoke(target, new[] {message});
+                }
+#elif ASYNC
+                if (TypeHelper.IsAssignableToGenericType(target.GetType(), typeof(IListenerAsync<>)))
+                {
+                    try
+                    {
+                        ((dynamic) _handlerMethod.Invoke(target, new[] {message})).Wait();

```

Comment:
This can cause deadlocks, a library should never .Wait() unless you take steps to remove the synchronisation context before invoking the handlers. If you cannot actually make this async, then use this:

```
((dynamic) Task.Factory.StartNew(() =>_handlerMethod.Invoke(target, new[] {message})).Wait(), TaskRunOptions.LongRunning); // I forget the enum type, but it gives it a dedicated thread which won't have a sync context
```